pkg/apis/autoscaling/types.go

```diff
@@ -420,3 +420,210 @@ type HorizontalPodAutoscalerList struct {
 	// Items is the list of horizontal pod autoscaler objects.
 	Items []HorizontalPodAutoscaler
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VerticalPodAutoscalerList is a list of VerticalPodAutoscaler objects.
+type VerticalPodAutoscalerList struct {
+	metav1.TypeMeta
+	// metadata is the standard list metadata.
+	// +optional
+	metav1.ListMeta
+
+	// items is the list of vertical pod autoscaler objects.
+	Items []VerticalPodAutoscaler
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VerticalPodAutoscaler is the configuration for a vertical pod
+// autoscaler, which automatically manages pod resources based on historical and
+// real time resource utilization.
+type VerticalPodAutoscaler struct {
+	metav1.TypeMeta
+	// Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the behavior of the autoscaler.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.
+	Spec VerticalPodAutoscalerSpec
+
+	// Current information about the autoscaler.
+	// +optional
+	Status VerticalPodAutoscalerStatus
+}
+
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec struct {
+	// A label query that determines the set of pods controlled by the Autoscaler.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector
```

Comment:
If this is not optional, should it be a pointer?  Looking at other users, it's always pointer=optional and non-pointer=required.  Which is it here?  It seems more common to define it as optional and have nil mean select all.

Reply:
A null label selector is defined to select no objects (https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors).
This doesn't make sense for VPA and is rejected at validation time. However I still want to make this field a pointer to force the user to set it explicitly, because an empty selector, while legal, has a very special meaning (match all) and is more likely to be user error than an intended use of the default.
I presume that other APIs followed similar reasoning, because I see many cases of the nil selector being rejected during validation, instead of making the field non-pointer.

Reply:
ok